storage/internal/benchmarks/continuous_reads.go

```diff
@@ -0,0 +1,202 @@
+	bucketName    string
+	directoryPath string
+	numWorkers    int
+	objects       chan string // list of objects to synchronize reads
+	results       []time.Duration
+	api           benchmarkAPI
+	objectSize    int64 // every object must be of the same size
+}
+
+func (r *continuousReads) setup(ctx context.Context) error {
+	// Check for context cancellation
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+	}
+
+	// Select API first as it will be the same for all writes/reads
+	api := opts.api
+	if api == mixedAPIs {
+		switch rand.Intn(4) {
+		case 0:
+			api = xmlAPI
+		case 1:
+			api = jsonAPI
+		case 2:
+			api = grpcAPI
+		case 3:
+			api = directPath
+		}
+	}
+	r.api = api
+
+	// Select object size
+	objectSize := opts.objectSize
+	if objectSize == 0 {
+		objectSize = randomInt64(opts.minObjectSize, opts.maxObjectSize)
+	}
+	r.objectSize = objectSize
+
+	// Make a temp dir for this run
+	dir, err := os.MkdirTemp("", "benchmark-experiment-")
+	if err != nil {
+		return err
+	}
+
+	// Create contents
+	r.directoryPath = dir
+
+	objects, err := generateDirInGCS(ctx, path.Base(dir), objectSize)
+	if err != nil {
+		return err
+	}
+	r.objects = *objects
+
+	return nil
+}
+
+// cleanup deletes objects on disk and in GCS. It does not accept a context as
+// it should run to completion to ensure full clean up of resources.
+func (r *continuousReads) cleanup() error {
+	// Clean temp dir
+	if err := os.RemoveAll(r.directoryPath); err != nil {
+		return err
+	}
+
+	// Delete uploaded objects
+	return deleteDirectoryFromGCS(r.bucketName, path.Base(r.directoryPath))
+}
+
+func (r *continuousReads) run(ctx context.Context) error {
+	benchGroup, ctx := errgroup.WithContext(ctx)
+	benchGroup.SetLimit(r.numWorkers)
+
+	for {
+		select {
+		case <-ctx.Done():
+			err := benchGroup.Wait()
+			r.compileResults()
+			if errorIsDeadLineExceeded(err) {
+				return nil
+			}
+			return err
+		default:
+			benchGroup.Go(func() error {
+				client := getClient(ctx, r.api)
+				object := <-r.objects
+
+				var span trace.Span
+				ctx, span = otel.GetTracerProvider().Tracer(tracerName).Start(ctx, "continuous_reads")
+				span.SetAttributes(attribute.KeyValue{"workload", attribute.StringValue("9")},
+					attribute.KeyValue{"api", attribute.StringValue(string(r.api))},
+					attribute.KeyValue{"object_size", attribute.Int64Value(r.objectSize)})
+				defer span.End()
+
+				// Download full object (range is not supported)
+				rangeStart := int64(0)
+				rangeLength := int64(-1)
+
+				timeTaken, err := downloadBenchmark(ctx, downloadOpts{
+					client:              client,
+					objectSize:          r.objectSize,
+					bucket:              r.bucketName,
+					object:              object,
+					rangeStart:          rangeStart,
+					rangeLength:         rangeLength,
+					downloadToDirectory: r.directoryPath,
+					timeout:             r.opts.timeoutPerOp,
+				})
+
+				if err != nil {
+					r.results = append(r.results, timeTaken)
```

Comment:
This seems potentially not thread safe? Do we need a mutex? Might be good to run it with the race detector to check.

Reply:
Yeah, you're right. I'll send it to a channel instead.